File: ios-native/ClubFuoco/Features/Explore/PersonalizationScore.swift
import Foundation

/// Native port of the web feed's personalisation layer — surveyScore(),
/// tasteScore() and prefScore() in src/app/(app)/explore/page.tsx (~165–268).
/// THE WEIGHTS MUST MATCH THE WEB'S so both platforms rank the same user
/// comparably — change them together or the two feeds drift apart.

/// The user's onboarding preferences (users.preferences JSON). Only the fields
/// prefScore reads; Codable ignores the rest.
struct UserPreferences: Decodable, Sendable {
    var budget: Double?
    var vibes: [String]?
    var crowd: String?
}

/// The survey-derived preference profile from GET /api/surveys/preferences —
/// the same payload the web feed consumes, derived once server-side
/// (deriveSurveyPreferences in src/lib/survey-preferences.ts), so both
/// platforms score identical signals by construction. Only the fields the
/// scorer reads are modelled; Codable ignores the rest. All optional so a
/// payload shape change degrades a signal to neutral instead of nil-ing the
/// whole profile.
struct SurveyPreferences: Decodable, Sendable {
    var surveyCount: Int?
    var avgRating: Double?
    var avgVibeRating: Double?
    var avgCrowdRating: Double?
    var preferredPriceLevel: Int?
    var likesCocktails: Bool?
    var likesBeer: Bool?
    var likesWine: Bool?
    var likesShots: Bool?
    var goodVibeAtClub: Bool?
    var goodVibeAtBar: Bool?
    var likesBusyVenues: Bool?
    var likedVenueNames: [String]?
    var avoidPlaceNames: [String]?
}

/// Row from user_taste_profiles (computed from bookings + surveys + tags).
struct TasteProfile: Decodable, Sendable {
    var topNeighborhoods: [String]?
    var topGenres: [String]?
    var topVibes: [String]?
}

enum PersonalizationScore {
    /// Web's anyHas(): keyword against name, address/neighbourhood, stored
    /// music genres and tags. (Place.matches() skips the address — the web
    /// scoring doesn't, so this stays its own helper.)
    private static func anyHas(_ p: Place, _ kws: [String]) -> Bool {
        let name = p.name.lowercased()
        let addr = (p.address + " " + (p.neighborhood ?? "")).lowercased()
        if kws.contains(where: { name.contains($0) || addr.contains($0) }) { return true }
        let genresAndTags = (p.musicGenres + p.tags).map { $0.lowercased() }
        return kws.contains { kw in genresAndTags.contains { $0.contains(kw) } }
    }

    /// Web's normV(): lowercase, strip non-alphanumerics.
    private static func normV(_ s: String) -> String {
        String(s.lowercased().unicodeScalars.filter { CharacterSet.alphanumerics.contains($0) })
    }

    static func budgetToPriceLevel(_ euros: Double) -> Int {
        if euros >= 999 { return 4 }
        if euros >= 80 { return 3 }
        if euros >= 40 { return 2 }
        if euros >= 20 { return 1 }
        return 0
    }

    /// Survey-based boost — same branches and weights as the web's surveyScore().
    static func surveyScore(_ p: Place, survey: SurveyPreferences?) -> Double {
        guard let survey else { return 0 }
        var score = 0.0
        let nameL = p.name.lowercased()
        let norm = normV(p.name)

        // Venues they loved → strong boost for similar (name-matched) venues
        for liked in survey.likedVenueNames ?? [] {
            if norm == normV(liked) || nameL.contains(liked.lowercased().prefix(6)) { score += 5 }
        }
        // Venues they hated → penalise
        for avoid in survey.avoidPlaceNames ?? [] {
            if norm == normV(avoid) || nameL.contains(avoid.lowercased().prefix(6)) { score -= 10 }
        }

        // Drink signal → venue type boost
        if survey.likesCocktails == true, anyHas(p, ["cocktail", "mixology", "speakeasy", "gin", "craft"]) { score += 3 }
        if survey.likesBeer == true, anyHas(p, ["pub", "cervecería", "brew", "beer", "craft"]) { score += 3 }
        if survey.likesWine == true, anyHas(p, ["wine", "bodega", "vinoteca", "vino"]) { score += 3 }
        if survey.likesShots == true, anyHas(p, ["club", "disco", "party", "night"]) { score += 2 }

        // Preferred price level
        if let preferred = survey.preferredPriceLevel, let level = p.priceLevel {
            score += Double(max(0, 3 - abs(level - preferred)))
        }

        // Vibe signals — positive only: they proved they enjoy this venue type's energy
        if survey.goodVibeAtClub == true, anyHas(p, ["club", "disco", "rave", "dance", "techno", "house", "sala"]) { score += 2 }
        if survey.goodVibeAtBar == true, anyHas(p, ["bar", "pub", "lounge", "cafe", "cocktail", "jazz", "wine"]) { score += 2 }
        if survey.likesBusyVenues == true, p.ratingsTotal > 200 || (p.rating ?? 0) >= 4.3 { score += 1 }

        return score
    }

    /// Taste-profile boost from stored tag-based signals — web's tasteScore().
    static func tasteScore(_ p: Place, taste: TasteProfile?) -> Double {
        guard let taste else { return 0 }
        var score = 0.0
        let nameL = p.name.lowercased()
        let addr = p.address.lowercased()

        // Neighbourhood match
        for n in taste.topNeighborhoods ?? [] {
            let nl = n.lowercased()
            if addr.contains(nl) || nameL.contains(nl) { score += 3 }
        }

        // Music genre tags → boost venues whose names hint at the genre
        let genreKw: [String: [String]] = [
            "techno":     ["techno", "input", "nitsa", "bunker", "sala"],
            "house":      ["house", "pacha", "bling"],
            "latin":      ["latin", "salsa", "reggaeton", "shoko", "latino"],
            "hip_hop":    ["sutton", "urban", "hip", "hop", "otto"],
            "indie":      ["indie", "apolo", "razzmatazz", "razz"],
            "electronic": ["electronic", "moog", "macarena", "mondo"],
            "jazz":       ["jazz", "blues", "acoustic", "live"],
        ]
        for genre in taste.topGenres ?? [] {
            let kws = genreKw[genre] ?? []
            if kws.contains(where: { nameL.contains($0) }) { score += 4 }
        }

        // Vibe tags
        let vibeKw: [String: [String]] = [
            "upscale":   ["club", "lounge", "vip", "sutton", "bling", "pacha"],
            "budget":    ["pub", "bar", "cervecería", "brew"],
            "rooftop":   ["rooftop", "terraza", "sky", "top"],
            "mid_range": ["cocktail", "bistro", "café"],
        ]
        for vibe in taste.topVibes ?? [] {
            let kws = vibeKw[vibe] ?? []
            if kws.contains(where: { nameL.contains($0) }) { score += 2 }
        }

        return score
    }

    /// The combined personalisation score — web's prefScore(): both signal
    /// layers plus onboarding budget/vibe/crowd preferences and rating bumps.
    static func prefScore(
        _ p: Place,
        prefs: UserPreferences?,
        survey: SurveyPreferences?,
        taste: TasteProfile?
    ) -> Double {
        var score = surveyScore(p, survey: survey) + tasteScore(p, taste: taste)
        guard let prefs else { return score }

        if let budget = prefs.budget, budget > 0, let level = p.priceLevel {
            let target = budgetToPriceLevel(budget)
            score += Double(max(0, 3 - abs(level - target)))
        }

        let nameL = p.name.lowercased()
        let vibeKw: [String: [String]] = [
            "beach":       ["beach", "mar", "maritim", "port"],
            "rooftop":     ["rooftop", "sky", "terraza", "top"],
            "upscale":     ["club", "lounge", "vip", "suite"],
            "underground": ["underground", "bunker", "basement", "techno"],
            "live_music":  ["music", "jazz", "live", "concert"],
            "wild":        ["club", "disco", "party"],
            "intimate":    ["bar", "bistro", "wine"],
        ]
        for v in prefs.vibes ?? [] {
            if vibeKw[v]?.contains(where: { nameL.contains($0) }) == true { score += 2 }
        }
        if prefs.crowd == "lgbtq" {
            if ["arena", "metro", "pride", "gay"].contains(where: { nameL.contains($0) }) { score += 3 }
        }
        if let rating = p.rating, rating >= 4.2 { score += 1 }
        if let rating = p.rating, rating >= 4.5 { score += 1 }
        return score
    }
}

File: android-native/app/src/main/java/com/clubfuoco/app/features/explore/PersonalizationScore.kt
package com.clubfuoco.app.features.explore

import com.clubfuoco.app.models.Place
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.math.abs

/** The user's onboarding preferences (users.preferences JSON). */
@Serializable
data class UserPreferences(
    val budget: Double? = null,
    val vibes: List<String>? = null,
    val crowd: String? = null,
)

/**
 * The survey-derived preference profile from GET /api/surveys/preferences — the
 * same payload the web feed consumes, derived once server-side so both
 * platforms score identical signals by construction.
 *
 * All optional so a payload shape change degrades a signal to neutral instead
 * of nulling the whole profile.
 */
@Serializable
data class SurveyPreferences(
    @SerialName("surveyCount") val surveyCount: Int? = null,
    @SerialName("avgRating") val avgRating: Double? = null,
    @SerialName("avgVibeRating") val avgVibeRating: Double? = null,
    @SerialName("avgCrowdRating") val avgCrowdRating: Double? = null,
    @SerialName("preferredPriceLevel") val preferredPriceLevel: Int? = null,
    @SerialName("likesCocktails") val likesCocktails: Boolean? = null,
    @SerialName("likesBeer") val likesBeer: Boolean? = null,
    @SerialName("likesWine") val likesWine: Boolean? = null,
    @SerialName("likesShots") val likesShots: Boolean? = null,
    @SerialName("goodVibeAtClub") val goodVibeAtClub: Boolean? = null,
    @SerialName("goodVibeAtBar") val goodVibeAtBar: Boolean? = null,
    @SerialName("likesBusyVenues") val likesBusyVenues: Boolean? = null,
    @SerialName("likedVenueNames") val likedVenueNames: List<String>? = null,
    @SerialName("avoidPlaceNames") val avoidPlaceNames: List<String>? = null,
)

/** Row from user_taste_profile (computed from bookings + surveys + tags). */
@Serializable
data class TasteProfile(
    @SerialName("top_neighborhoods") val topNeighborhoods: List<String>? = null,
    @SerialName("top_genres") val topGenres: List<String>? = null,
    @SerialName("top_vibes") val topVibes: List<String>? = null,
)

/**
 * Port of the web feed's personalisation layer — surveyScore(), tasteScore()
 * and prefScore().
 *
 * THE WEIGHTS MUST MATCH THE WEB AND iOS so all platforms rank the same user
 * comparably. Change them together.
 */
object PersonalizationScore {

    /**
     * Web's anyHas(): keyword against name, address/neighbourhood, stored music
     * genres and tags. (`Place.matches` skips the address — the web scoring
     * does not, so this stays its own helper.)
     */
    private fun anyHas(p: Place, kws: List<String>): Boolean {
        val name = p.name.lowercase()
        val addr = (p.address + " " + (p.neighborhood ?: "")).lowercase()
        if (kws.any { name.contains(it) || addr.contains(it) }) return true
        val genresAndTags = (p.musicGenres + p.tags).map { it.lowercase() }
        return kws.any { kw -> genresAndTags.any { it.contains(kw) } }
    }

    /** Web's normV(): lowercase, strip non-alphanumerics. */
    private fun normV(s: String): String =
        s.lowercase().filter { it.isLetterOrDigit() }

    fun budgetToPriceLevel(euros: Double): Int = when {
        euros >= 999 -> 4
        euros >= 80 -> 3
        euros >= 40 -> 2
        euros >= 20 -> 1
        else -> 0
    }

    /** Survey-based boost — same branches and weights as the web's surveyScore(). */
    fun surveyScore(p: Place, survey: SurveyPreferences?): Double {
        if (survey == null) return 0.0
        var score = 0.0
        val nameL = p.name.lowercase()
        val norm = normV(p.name)

        // Venues they loved → strong boost for similar (name-matched) venues.
        for (liked in survey.likedVenueNames.orEmpty()) {
            if (norm == normV(liked) || nameL.contains(liked.lowercase().take(6))) score += 5
        }
        // Venues they hated → penalise.
        for (avoid in survey.avoidPlaceNames.orEmpty()) {
            if (norm == normV(avoid) || nameL.contains(avoid.lowercase().take(6))) score -= 10
        }

        // Drink signal → venue type boost.
        if (survey.likesCocktails == true &&
            anyHas(p, listOf("cocktail", "mixology", "speakeasy", "gin", "craft"))
        ) score += 3
        if (survey.likesBeer == true &&
            anyHas(p, listOf("pub", "cervecería", "brew", "beer", "craft"))
        ) score += 3
        if (survey.likesWine == true &&
            anyHas(p, listOf("wine", "bodega", "vinoteca", "vino"))
        ) score += 3
        if (survey.likesShots == true &&
            anyHas(p, listOf("club", "disco", "party", "night"))
        ) score += 2

        // Preferred price level.
        val preferred = survey.preferredPriceLevel
        val level = p.priceLevel
        if (preferred != null && level != null) {
            score += maxOf(0, 3 - abs(level - preferred)).toDouble()
        }

        // Vibe signals — positive only: they proved they enjoy this energy.
        if (survey.goodVibeAtClub == true &&
            anyHas(p, listOf("club", "disco", "rave", "dance", "techno", "house", "sala"))
        ) score += 2
        if (survey.goodVibeAtBar == true &&
            anyHas(p, listOf("bar", "pub", "lounge", "cafe", "cocktail", "jazz", "wine"))
        ) score += 2
        if (survey.likesBusyVenues == true &&
            (p.ratingsTotal > 200 || (p.rating ?: 0.0) >= 4.3)
        ) score += 1

        return score
    }

    /** Taste-profile boost from stored tag-based signals — web's tasteScore(). */
    fun tasteScore(p: Place, taste: TasteProfile?): Double {
        if (taste == null) return 0.0
        var score = 0.0
        val nameL = p.name.lowercase()
        val addr = p.address.lowercase()

        for (n in taste.topNeighborhoods.orEmpty()) {
            val nl = n.lowercase()
            if (addr.contains(nl) || nameL.contains(nl)) score += 3
        }

        val genreKw = mapOf(
            "techno" to listOf("techno", "input", "nitsa", "bunker", "sala"),
            "house" to listOf("house", "pacha", "bling"),
            "latin" to listOf("latin", "salsa", "reggaeton", "shoko", "latino"),
            "hip_hop" to listOf("sutton", "urban", "hip", "hop", "otto"),
            "indie" to listOf("indie", "apolo", "razzmatazz", "razz"),
            "electronic" to listOf("electronic", "moog", "macarena", "mondo"),
            "jazz" to listOf("jazz", "blues", "acoustic", "live"),
        )
        for (genre in taste.topGenres.orEmpty()) {
            if (genreKw[genre].orEmpty().any { nameL.contains(it) }) score += 4
        }

        val vibeKw = mapOf(
            "upscale" to listOf("club", "lounge", "vip", "sutton", "bling", "pacha"),
            "budget" to listOf("pub", "bar", "cervecería", "brew"),
            "rooftop" to listOf("rooftop", "terraza", "sky", "top"),
            "mid_range" to listOf("cocktail", "bistro", "café"),
        )
        for (vibe in taste.topVibes.orEmpty()) {
            if (vibeKw[vibe].orEmpty().any { nameL.contains(it) }) score += 2
        }

        return score
    }

    /**
     * The combined personalisation score — web's prefScore(): both signal
     * layers plus onboarding budget/vibe/crowd preferences and rating bumps.
     */
    fun prefScore(
        p: Place,
        prefs: UserPreferences?,
        survey: SurveyPreferences?,
        taste: TasteProfile?,
    ): Double {
        var score = surveyScore(p, survey) + tasteScore(p, taste)
        if (prefs == null) return score

        val budget = prefs.budget
        val level = p.priceLevel
        if (budget != null && budget > 0 && level != null) {
            val target = budgetToPriceLevel(budget)
            score += maxOf(0, 3 - abs(level - target)).toDouble()
        }

        val nameL = p.name.lowercase()
        val vibeKw = mapOf(
            "beach" to listOf("beach", "mar", "maritim", "port"),
            "rooftop" to listOf("rooftop", "sky", "terraza", "top"),
            "upscale" to listOf("club", "lounge", "vip", "suite"),
            "underground" to listOf("underground", "bunker", "basement", "techno"),
            "live_music" to listOf("music", "jazz", "live", "concert"),
            "wild" to listOf("club", "disco", "party"),
            "intimate" to listOf("bar", "bistro", "wine"),
        )
        for (v in prefs.vibes.orEmpty()) {
            if (vibeKw[v]?.any { nameL.contains(it) } == true) score += 2
        }
        if (prefs.crowd == "lgbtq" &&
            listOf("arena", "metro", "pride", "gay").any { nameL.contains(it) }
        ) score += 3

        val rating = p.rating
        if (rating != null && rating >= 4.2) score += 1
        if (rating != null && rating >= 4.5) score += 1
        return score
    }
}
